src/invidious/user/imports.cr

```diff
@@ -290,42 +290,38 @@ struct Invidious::User
     end
 
     def from_newpipe(user : User, body : String) : Bool
-      io = IO::Memory.new(body)
-
-      Compress::Zip::File.open(io) do |file|
+      Compress::Zip::File.open(IO::Memory.new(body), true) do |file|
         file.entries.each do |entry|
           entry.open do |file_io|
-            # Ensure max size of 4MB
-            io_sized = IO::Sized.new(file_io, 0x400000)
-
             next if entry.filename != "newpipe.db"
 
-            tempfile = File.tempfile(".db")
-
-            begin
-              File.write(tempfile.path, io_sized.gets_to_end)
-            rescue
-              return false
-            end
-
-            db = DB.open("sqlite3://" + tempfile.path)
+            # Ensure max size of 4MB
+            io_sized = IO::Sized.new(file_io, 0x400000)
 
-            user.watched += db.query_all("SELECT url FROM streams", as: String)
-              .map(&.lchop("https://www.youtube.com/watch?v="))
+            temp = File.tempfile(".db") do |tempfile|
+              begin
+                File.write(tempfile.path, io_sized.gets_to_end)
+              rescue
+                return false
+              end
 
-            user.watched.uniq!
-            Invidious::Database::Users.update_watch_history(user)
+              DB.open("sqlite3://" + tempfile.path) do |db|
+                user.watched += db.query_all("SELECT url FROM streams", as: String)
+                  .map(&.lchop("https://www.youtube.com/watch?v="))
 
-            user.subscriptions += db.query_all("SELECT url FROM subscriptions", as: String)
-              .map(&.lchop("https://www.youtube.com/channel/"))
+                user.watched.uniq!
+                Invidious::Database::Users.update_watch_history(user)
 
-            user.subscriptions.uniq!
-            user.subscriptions = get_batch_channels(user.subscriptions)
+                user.subscriptions += db.query_all("SELECT url FROM subscriptions", as: String)
+                  .map(&.lchop("https://www.youtube.com/channel/"))
 
-            Invidious::Database::Users.update_subscriptions(user)
+                user.subscriptions.uniq!
+                user.subscriptions = get_batch_channels(user.subscriptions)
 
-            db.close
-            tempfile.delete
+                Invidious::Database::Users.update_subscriptions(user)
+              end
+            end
+            temp.delete
```

Comment:
I realized that the file will not be deleted in the case of a database error (i.e if `DB.open` raises an exception). You'll need to wrap the entire thing in a `begin` / (`rescue`) / `ensure`.

Reply:
Done :smile: